File: sections/27_bitmask_dp/problems/a_skill_coverage/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    it = iter(data)
    n = next(it)
    m = next(it)
    full = (1 << n) - 1
    cheapest_for_mask: dict[int, int] = {}
    for _ in range(m):
        cost = next(it)
        k = next(it)
        mask = 0
        for _ in range(k):
            skill = next(it) - 1
            mask |= 1 << skill
        previous = cheapest_for_mask.get(mask)
        if previous is None or cost < previous:
            cheapest_for_mask[mask] = cost

    # An offer that costs no less and covers a subset of another offer can
    # never appear in an optimum. Removing it keeps the worst-case DP within
    # the exercise's Python time budget.
    raw_offers = list(cheapest_for_mask.items())
    offers = [
        (cost, mask)
        for mask, cost in raw_offers
        if not any(
            other_mask != mask
            and (mask | other_mask) == other_mask
            and other_cost <= cost
            for other_mask, other_cost in raw_offers
        )
    ]

    inf = 10**30
    dp = [inf] * (1 << n)
    dp[0] = 0
    for cost, offer in offers:
        new_dp = dp[:]
        for mask, value in enumerate(dp):
            if value == inf:
                continue
            candidate = value + cost
            merged = mask | offer
            if candidate < new_dp[merged]:
                new_dp[merged] = candidate
        dp = new_dp

    print(-1 if dp[full] >= inf else dp[full])


if __name__ == "__main__":
    main()

File: sections/27_bitmask_dp/problems/a_skill_coverage/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    if (!(cin >> n >> m)) return 0;
    vector<pair<long long, int>> offers;
    for (int i = 0; i < m; i++) {
        long long cost;
        int k;
        cin >> cost >> k;
        int mask = 0;
        for (int j = 0; j < k; j++) {
            int skill;
            cin >> skill;
            mask |= 1 << (skill - 1);
        }
        offers.push_back({cost, mask});
    }

    int total = 1 << n;
    int full = total - 1;
    const long long INF = (1LL << 62);
    vector<long long> dp(total, INF);
    dp[0] = 0;

    for (auto [cost, offer] : offers) {
        vector<long long> next_dp = dp;
        for (int mask = 0; mask < total; mask++) {
            if (dp[mask] == INF) continue;
            int merged = mask | offer;
            next_dp[merged] = min(next_dp[merged], dp[mask] + cost);
        }
        dp.swap(next_dp);
    }

    cout << (dp[full] == INF ? -1 : dp[full]) << '\n';
    return 0;
}
